nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-components/src/main/java/org/apache/nifi/controller/repository/StandardProcessSession.java

```diff
@@ -356,6 +408,117 @@ private void checkpoint(final boolean copyCollections) {
         checkpoint.checkpoint(this, autoTerminatedEvents, copyCollections);
     }
 
+    private boolean isRetryNeeded(final Connectable connectable, final StandardRepositoryRecord record, final FlowFileRecord currentFlowFile,
+                                  final int retryCount, final Map<String, StandardRepositoryRecord> uuidsToRecords) {
+        if (currentFlowFile == null || connectable == null || connectable.getRetriedRelationships().isEmpty()) {
+            return false;
+        }
+
+        if (connectable.isRelationshipRetried(record.getTransferRelationship())) {
+            return retryCount < connectable.getRetryCount();
+        }
+
+        final ProvenanceEventBuilder eventBuilder = forkEventBuilders.get(currentFlowFile);
+        if (eventBuilder != null) {
+            for (String uuid : eventBuilder.getChildFlowFileIds()) {
+                if (connectable.isRelationshipRetried(uuidsToRecords.get(uuid).getTransferRelationship())) {
+                    return retryCount < connectable.getRetryCount();
+                }
+            }
+        }
```

Comment:
I'm a bit concerned about the use of provenance events here for determining whether or not a FlowFile needs to be retried. I can definitely appreciate why that association was made here - it's the data model that we have that also encapsulates this linkage. But the provenance is really about "auditing" moreso than the logic of how FlowFiles within the session should be linked. This also can get very expensive if we are dealing with multiple splits. It probably makes sense to introduce a separate linkage between the FlowFiles outside of the provenance events.